vm/src/rv32/parse.rs

```diff
@@ -162,18 +180,28 @@ mod opcodes {
 }
 pub use opcodes::*;
 
-// parse a 32-bit word as an instruction
+fn assert_option(exp: bool) -> Option<()> {
+    if exp {
+        Some(())
+    } else {
+        None
+    }
+}
 
+// parse a 32-bit word as an instruction
 pub(crate) fn parse_u32(word: u32) -> Option<RV32> {
     let inst = match opcode(word) {
         OPC_LUI => LUI { rd: rd(word), imm: immU(word) },
         OPC_AUIPC => AUIPC { rd: rd(word), imm: immU(word) },
         OPC_JAL => JAL { rd: rd(word), imm: immJ(word) },
-        OPC_JALR => JALR {
-            rd: rd(word),
-            rs1: rs1(word),
-            imm: immI(word),
-        },
+        OPC_JALR => {
+            assert_option(funct3(word) == 0)?;
```

Comment:
Is there a reason this doesn't panic? We can generally panic (`debug_assert!`) within this circuit construction code: for deployment we only ever construct a single step circuit using it, so it'll be caught in CI testing if anything blows up.

Reply:
`parse_u32` was already using `Option` to catch instruction formatting errors. I just reused the path.

Reply:
Ah I missed that.